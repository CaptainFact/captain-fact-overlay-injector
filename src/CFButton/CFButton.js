import 'isomorphic-fetch'
import React from 'react'
import { connect } from 'react-redux'
import classnames from 'classnames'

import { cfbutton, pulse } from './CFButton.css'
import { InterfaceState } from '../App/interface_reducer'
import { getFocusedStatement } from '../Statement/selectors'
import { ICONS } from '../config'


@connect(state => ({
  statements: state.Statements.data,
  displayed: state.Interface.sidebarCollapsed,
  video: state.Video.data,
  statement: getFocusedStatement(state)
}))
export default class CFButton extends React.Component {
  getIcon() {
    if (!this.props.statement)
      return ICONS.neutral

    const globalScore = this.props.statement.comments.reduce((score, comment) =>
      score + (comment.approve ? comment.score : -comment.score)
    , 0)
    if (globalScore > 0)
      return ICONS.confirm
    else if (globalScore < 0)
      return ICONS.refute
    else
      return ICONS.neutral
  }

  render() {
    if (!this.props.video || this.props.statements.size === 0)
      return null
    return (
      <img src={this.getIcon()}
        className={classnames("CFButton", cfbutton, {[pulse]: !!this.props.statement})}
        title="CaptainFact"
        onClick={InterfaceState.openSidebar}
        style={{
          display: this.props.displayed ? "block" : "none"
        }}
      />
    )
  }
}