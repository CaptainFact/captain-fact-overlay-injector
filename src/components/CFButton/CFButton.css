.cfbutton {
  font-size: 1em;
  height: 2.75em;
  min-width: 40px;
  max-width: 60px;
  min-height: 40px;
  max-height: 60px;
  position: absolute;
  right: 1.7em;
  top: 3.5em;
  z-index: 29;
  cursor: pointer;
  box-shadow: 1px 1px 2px 0 rgba(0, 0, 0, 0.60);
  border-radius: 2em;
  display: block;
}

.cfbutton.hidden {
  display: none;
}

.cfbutton:hover {
  transform: scale(1.1);
}

.cfbutton:not(.pulse):not(:hover) {
  opacity: 0.4;
}

.pulse:not(:hover) {
  animation-name: pulse;
  animation-duration: 3000ms;
  transform-origin: 50% 50%;
  animation-iteration-count: infinite;
  animation-timing-function: linear;
}

@keyframes pulse {
  0%    { transform: rotate(0deg)   scale(1); box-shadow: 1px 1px 2px 0 rgba(0, 0, 0, 0.60); }
  10%   { transform: rotate(0deg)   scale(1); box-shadow: 1px 1px 2px 0 rgba(0, 0, 0, 0.60); }
  14%   { transform: rotate(-10deg) scale(1.4); box-shadow: 1px 1px 0 0 rgba(0, 0, 0, 0.60); }
  20%   { transform: rotate(10deg)  scale(1); box-shadow: 1px 1px 2px 0 rgba(0, 0, 0, 0.60); }
  23%   { transform: rotate(5deg)   scale(1.2); box-shadow: 1px 1px 1px 0 rgba(0, 0, 0, 0.60); }
  26%   { transform: rotate(0deg)   scale(1); box-shadow: 1px 1px 2px 0 rgba(0, 0, 0, 0.60); }
  100%  { transform: rotate(0deg)   scale(1); box-shadow: 1px 1px 2px 0 rgba(0, 0, 0, 0.60); }
}