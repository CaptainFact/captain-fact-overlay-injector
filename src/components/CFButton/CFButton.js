import 'isomorphic-fetch'
import React from 'react'
import { connect } from 'react-redux'
import classnames from 'classnames'
import { translate } from 'react-i18next'

import { cfbutton, pulse, hidden } from './CFButton.css'
import { InterfaceState } from '../App/interface_reducer'
import { getFocusedStatement } from '../Statement/selectors'
import { getGraphics } from '../App/Configuration/selectors'

import iconNeutral from '../../assets/icon.png'
import iconConfirm from '../../assets/icon_confirm.png'
import iconRefute from '../../assets/icon_refute.png'

@translate(['translations'], { wait: true })
export class CFButton extends React.PureComponent {
  render() {
    if (!this.props.hasVideo || !this.props.hasStatements)
      return null

    const globalScore = this.calculateGlobalScore()
    return (
      <img 
        src={this.getIcon(globalScore)}
        className={this.getClassNames()}
        title={ this.props.t('CaptainFact') }
        onClick={InterfaceState.openSidebar}
        alt={ this.props.t('CF') }
      />
    )
  }

  getClassNames() {
    return classnames(cfbutton, {
      [pulse]: !!this.props.statement,
      [hidden]: !this.props.displayed
    })
  }

  getIcon(globalScore) {
    if (globalScore > 0)
      return (this.props.icons && this.props.icons.confirm) || iconConfirm
    else if (globalScore < 0)
      return (this.props.icons && this.props.icons.refute) || iconRefute
    return (this.props.icons && this.props.icons.neutral) || iconNeutral
  }

  calculateGlobalScore() {
    // TODO This should be in Redux
    if (!this.props.statement)
      return 0

    return this.props.statement.comments.reduce((score, comment) => {
      if (comment.approve === true)
        return score + Math.max(comment.score, 0)
      else if (comment.approve === false)
        return score - Math.max(comment.score, 0)
      return score
    }, 0)
  }
}

export default connect(state => ({
  hasStatements: state.Statements.data.size !== 0,
  hasVideo: !!state.Video.data,
  displayed: state.Interface.sidebarCollapsed,
  statement: getFocusedStatement(state),
  icons: getGraphics(state).logo,
  baseSize: state.Configuration.app.baseSize
}))(CFButton)
