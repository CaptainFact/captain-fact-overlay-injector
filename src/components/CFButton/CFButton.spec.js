import React from 'react'
import { Record } from 'immutable'
import { CFButton } from './CFButton'


const VALID_PROPS = new Record({
  hasVideo: true,
  hasStatements: true,
  displayed: true,
  statement: {comments: []},
  icons: {
    neutral: 'neutral-icon.jpg',
    confirm: 'approve-icon.jpg',
    refute: 'refute-icon.jpg',
  }
})()

const buildProps = (customProps={}) => VALID_PROPS.mergeDeep(customProps).toJS()

const POSITIVE_PROPS = buildProps({statement: {comments: [
  {approve: true, score: 25},
  {approve: false, score: 10},
  {approve: true, score: 2},
]}})

const NEGATIVE_PROPS = buildProps({statement: {comments: [
  {approve: false, score: 25},
  {approve: false, score: 10},
  {approve: true, score: 2},
]}})

const NEUTRAL_PROPS = buildProps({statement: {comments: [
  {approve: false, score: 10},
  {approve: true, score: 10},
]}})

test('Returns nothing if it has no video or statement', () => {
  snapshot(<CFButton {...buildProps({hasVideo: false})}/>)
  snapshot(<CFButton {...buildProps({hasStatements: false})}/>)
})

test('neutral icon if unsure', () => {
  snapshot(<CFButton {...NEUTRAL_PROPS}/>)
})

test('approve icon if approve', () => {
  snapshot(<CFButton {...POSITIVE_PROPS}/>)
})

test('refute icon if refute', () => {
  snapshot(<CFButton {...NEGATIVE_PROPS}/>)
})

test('Has default icons', () => {
  // Without icons
  snapshot(<CFButton {...buildProps({icons: null})}/>)

  // With partial icons object
  snapshot(<CFButton {...buildProps({icons: {approved: 'APPROVED.JPG'}})}/>)
})
