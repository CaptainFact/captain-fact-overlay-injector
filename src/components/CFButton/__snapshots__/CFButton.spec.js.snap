// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Has default icons 1`] = `
<img
  alt="acronymCF"
  className="cfbutton pulse"
  onClick={[Function]}
  src="test-file-stub"
  title="CaptainFact"
/>
`;

exports[`Has default icons 2`] = `
<img
  alt="acronymCF"
  className="cfbutton pulse"
  onClick={[Function]}
  src="test-file-stub"
  title="CaptainFact"
/>
`;

exports[`Returns nothing if it has no video or statement 1`] = `""`;

exports[`Returns nothing if it has no video or statement 2`] = `""`;

exports[`approve icon if approve 1`] = `
<img
  alt="acronymCF"
  className="cfbutton pulse"
  onClick={[Function]}
  src="approve-icon.jpg"
  title="CaptainFact"
/>
`;

exports[`neutral icon if unsure 1`] = `
<img
  alt="acronymCF"
  className="cfbutton pulse"
  onClick={[Function]}
  src="neutral-icon.jpg"
  title="CaptainFact"
/>
`;

exports[`refute icon if refute 1`] = `
<img
  alt="acronymCF"
  className="cfbutton pulse"
  onClick={[Function]}
  src="refute-icon.jpg"
  title="CaptainFact"
/>
`;
